doc/developer_guide/psy_data.rst

```diff
@@ -550,6 +551,286 @@ and provides these two lists to the ``gen_code()`` function of its base class,
+    course also possible to ignore ``is_enabled`` and use a different
+    mechanism (e.g. the NVIDIA profiling wrapper library calls corresponding
+    start and stop functions in the NVIDIA profiler).
+
+Jinja Support:
+    The base class is creating using the template language jinja. It is
+    therefore easy to automatically create the base functions for the
+    argument types actually required by the wrapper library. See
+    :ref:`jinja` for details.
+
+
+.. _jinja:
+
+Jinja Support in the Base Class
++++++++++++++++++++++++++++++++
+The base class is contained in ``lib/psy_data_base.jinja``. It is processed
+with the script ``process.py``, which will print the processed file to
+stdout. The ``Makefile`` will automatically create the file
+``psy_data_base.f90`` from the jinja template and compile it. For convenience
+if jinja is not installed on a system, a processed version of
+``psy_data_base.f90`` is included in ``lib``. If you use the base class in
+a wrapper library, you have to process the template in your library directory
+with additional parameters to specify the required types and the prefix.
+Besides the name of the template file to process the ``process.py`` script
+takes the following parameters:
+
+-types:
+    A comma-separated list of Fortran basic types (no spaces allowed).
+    Supported are the type names:
+
+    ``real``:
+        32-bit floating point value
+    ``double``:
+        64-bit floating point value
+    ``int``:
+        32-bit integer value
+    ``long``:
+        64-bit integer value
+
+    Default value is ``real,double,int``.
+
+-dims:
+    A comma-separated list of dimensions (no spaced allowed). Default
+    value is ``1,2,3,4``.
+
+-prefix:
+    The prefix to use for the PSyData type and functions. Default is
+    emtpy (i.e. no prefix). If you specify a prefix, you have to
+    add the ``_`` between the prefix and name explicitly.
+
+For each specified type name the jinja template will create methods called
+``DeclareScalar{{type}}`` and ``ProvideScalar{{type}}`` for handling
+scalar parameters. For array parameters, the functions
+``DeclareArray{{dim}}d{{type}}`` and ``ProvideArray{{dim}}d{{type}}``
+will be created for each type and each specified number of dimensions.
+
+Below an example of using the ``process.py`` script in a Makefile for a read-only
+verification library (taken from ``lib/read_only/lfric/Makefile``):
+
+.. code-block:: Makefile
+
+    PROCESS_ARGS = -prefix=read_only_verification -types=real,int,double \
+                   -dims=1,2,3,4
+
+    psy_data_base.f90: ../../psy_data_base.jinja
+        ../../process.py $(PROCESS_ARGS) $< > psy_data_base.f90
+
+This will create the processed file ``psy_data_base.f90`` in the directory
+of the library (which has the advantage that you will be using consistent
+compiler settings in your library, and consistent parameter specifying
+the required types and dimensions). You still need to declare all these
```

Comment:
Perhaps add a reference to either the lfric or gocean read-only implementation as an example (if that makes sense)?

Reply:
I moved that paragraph to a better place, and documented it a bit better. But yes, you somehow have to list all the automatically created functions in one generic interface. This can be done either 'manually' (e.g. dl_esm_inf readonly), or via a jinja script. I've improved the docs to point these two approaches out.
I also believe that just above the paragraph you pointed I already state where the example code comes from, so I don't think we need another reference here (since this paragraph actually explains the very same code).